crates/pallet-receipts/src/lib.rs

```diff
@@ -0,0 +1,366 @@
+    /// The parent execution receipt is missing.
+    MissingParent,
+}
+
+impl<T: Config> Pallet<T> {
+    /// Returns the block number of the latest receipt.
+    pub fn head_receipt_number(domain_id: DomainId) -> T::BlockNumber {
+        let (_, best_number) = <ReceiptHead<T>>::get(domain_id);
+        best_number
+    }
+
+    /// Returns the block number of the oldest receipt still being tracked in the state.
+    pub fn oldest_receipt_number(domain_id: DomainId) -> T::BlockNumber {
+        Self::finalized_receipt_number(domain_id) + One::one()
+    }
+
+    /// Returns the block number of latest _finalized_ receipt.
+    pub fn finalized_receipt_number(domain_id: DomainId) -> T::BlockNumber {
+        let (_, best_number) = <ReceiptHead<T>>::get(domain_id);
+        best_number.saturating_sub(T::ReceiptsPruningDepth::get())
+    }
+
+    /// Initialize the genesis execution receipt
+    pub fn initialize_genesis_receipt(domain_id: DomainId, genesis_hash: T::Hash) {
+        let genesis_receipt = ExecutionReceipt {
+            primary_number: Zero::zero(),
+            primary_hash: genesis_hash,
+            domain_hash: T::DomainHash::default(),
+            trace: Vec::new(),
+            trace_root: Default::default(),
+        };
+        Self::apply_new_best_receipt(domain_id, &genesis_receipt);
+        // Explicitly initialize the oldest receipt number even not necessary as ValueQuery is used.
+        <OldestReceiptNumber<T>>::insert::<_, T::BlockNumber>(domain_id, Zero::zero());
+    }
+
+    /// Track the execution receipts for the domain
+    pub fn track_receipts(
+        domain_id: DomainId,
+        receipts: &[ExecutionReceipt<T::BlockNumber, T::Hash, T::DomainHash>],
+    ) -> Result<(), Error> {
+        let oldest_receipt_number = <OldestReceiptNumber<T>>::get(domain_id);
+        let (_, mut best_number) = <ReceiptHead<T>>::get(domain_id);
+
+        for receipt in receipts {
+            // Ignore the receipt if it has already been pruned.
+            if receipt.primary_number < oldest_receipt_number {
+                continue;
+            }
+
+            if receipt.primary_number <= best_number {
+                // Either increase the vote for a known receipt or add a fork receipt at this height.
+                Self::apply_non_new_best_receipt(domain_id, receipt);
+            } else if receipt.primary_number == best_number + One::one() {
+                Self::apply_new_best_receipt(domain_id, receipt);
+                Self::remove_expired_receipts(domain_id, receipt.primary_number);
+                best_number += One::one();
+            } else {
+                // Reject the entire Bundle due to the missing receipt(s) between [best_number, .., receipt.primary_number].
+                return Err(Error::MissingParent);
+            }
+        }
+        Ok(())
+    }
+
+    /// Submit fraud proof that targetted a given domain
+    pub fn submit_fraud_proof(domain_id: DomainId, fraud_proof: FraudProof) {
```

Comment:
Something `process_fraud_proof` is somehow nicer to me so that it's distinctive from the Call with same name in pallet-domains.